### ISSUE TITLE
New/replacement nodes stuck JOINING perpetually

### ISSUE THREAD

**Issue by powellchristoph** [scylladb/scylla#7476]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [X] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 4.1.8-0.20201002.6c43a0dc29-1
Cluster size: 24 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 1604

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): AWS i3.4xlarge
Hardware: sockets= cores= hyperthreading= memory=
Disks: (SSD/HDD, count)

I am attempting to add additional nodes to my cluster and they are continually stuck in the `JOINING` state. While attempting to add nodes, I had a node failure. Attempts to replace that node according to the documentation has resulted in the same problem.

There have not been any active `stream_session`s in the logs for 24 hours. There are no errors or logs. The node appear stuck.
```
Mode: JOINING
Bootstrap-scorecard-index-8 c42f84c0-13bc-11eb-8ff4-000000000003
    /3.82.217.32
        Receiving 1 files, 0 bytes total. Already received 1 files, 61885535031 bytes total
            rxnofile 61885535031/61885535031 bytes(100%) received from idx:0/3.82.217.32
Bootstrap-scorecard-index-9 8b9764b0-13c2-11eb-8ff4-000000000003
    /54.157.198.151
        Receiving 1 files, 0 bytes total. Already received 1 files, 59677177829 bytes total
            rxnofile 59677177829/59677177829 bytes(100%) received from idx:0/54.157.198.151
Read Repair Statistics:
Attempted: 0
Mismatch (Blocking): 0
Mismatch (Background): 0
Pool Name                    Active   Pending      Completed   Dropped
Large messages                  n/a         1      768130040         0
Small messages                  n/a         0              0         0
Gossip messages                 n/a         0              0         0
```

I have stopped the joining nodes, removed their data and restarted the join. They all get stuck.

**Comment by powellchristoph** [comment]
Logs from scylla-scorecard-n15.use1. It was logging in debug mode, sorry for the size.
https://drive.google.com/file/d/1r6mMXwOa_-g8FZEI5DdFBJu0sQ1d3YT3/view?usp=sharing

scylla-scorecard-n01.use1
https://drive.google.com/file/d/1xUlEdHiJfKX3HbJ1TDhZ_qcgY6Jb418v/view?usp=sharing

**Comment by powellchristoph** [comment]
The node is now stuck without any active streams.

```
[prod-us-east-1]cpowell@scylla-scorecard-a-n15:~$ nodetool netstats
Mode: JOINING
Not sending any streams.
Read Repair Statistics:
Attempted: 0
Mismatch (Blocking): 0
Mismatch (Background): 0
Pool Name                    Active   Pending      Completed   Dropped
Large messages                  n/a         0      363666529         0
Small messages                  n/a         0              0         0
Gossip messages                 n/a         0              0         0
```

**Comment by slivne** [comment]
@powellchristoph we do not have access to those files (can you provide access for view for anyone in scylladb.com - in the future please use the documented method https://docs.scylladb.com/troubleshooting/report_scylla_problem/ )

**Comment by asias** [comment]
> I am attempting to add additional nodes to my cluster and they are continually stuck in the `JOINING` state. While attempting to add nodes, I had a node failure. Attempts to replace that node according to the documentation has resulted in the same problem.

So you performed a replace node operation (using --replace-address) while the bootstrap operation is in progress? Do you have MV tables  / large paritiotns / large rows.  The stuck starts to happen before or after the replace operation starts.

**Comment by asias** [comment]
Did all the nodes in the cluster use 4.1.8? Please make sure all of them use the same and latest 4.1.8 version.

**Comment by powellchristoph** [comment]
Hi @asias , sorry for the permission issue and slow response, I was out sick. You should be able to get the logs now. All nodes have been updated and are running the most recent release. I do not have any MV tables, but there is a potential for large rows.

Also, the node is still. stuck joining.
```
[prod-us-east-1]cpowell@scylla-scorecard-a-n15:~$ nodetool netstats
Mode: JOINING
Not sending any streams.
Read Repair Statistics:
Attempted: 0
Mismatch (Blocking): 0
Mismatch (Background): 0
Pool Name                    Active   Pending      Completed   Dropped
Large messages                  n/a         0      363666529         0
Small messages                  n/a         0              0         0
Gossip messages                 n/a         0              0         0
```

**Comment by asias** [comment]
@powellchristoph Thanks. 

The joining node does not stuck in streaming. The streaming for bootstrap failed and it was retried. 

```
1033377 Oct 22 19:14:18 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839aa000-149a-11eb-8ff4-000000000003] Received failed complete message, peer=3.82.217.32
1033378 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839aa000-149a-11eb-8ff4-000000000003] Streaming plan for Bootstrap-scorecard-index-0 failed, peers={3.82.217.32},         tx=0 KiB, 0.00 KiB/s, rx=131015 KiB, 1131.17 KiB/s
1033379 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] rpc - client 10.2.96.53:7001: client connection dropped: read: Connection reset by peer
1033380 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839bd880-149a-11eb-8ff4-000000000003] Received failed complete message, peer=54.157.198.151
1033381 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap with 3.82.217.32 for keyspace=scorecard failed, took 116.496 seconds: streaming::stream_exception (Stream failed)
1033382 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839bd880-149a-11eb-8ff4-000000000003] Streaming plan for Bootstrap-scorecard-index-0 failed, peers={54.157.198.151        }, tx=0 KiB, 0.00 KiB/s, rx=136129 KiB, 1170.94 KiB/s
1033383 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap with 54.157.198.151 for keyspace=scorecard failed, took 116.957 seconds: streaming::stream_exception (Stream fail        ed)
1033384 Oct 22 19:14:20 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap failed, took 134 seconds, nr_ranges_remaining=1377
1033385 Oct 22 19:14:20 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap failed to stream. Will retry in 90 seconds ...

```


So n15(52.55.159.2) is the bootstrapping node. On which node did you run replace operation? You mentioned you did replace operation. Who were node  54.157.198.151 / 52.26.187.180 and 3.82.217.32? What happened to them?

```

n1:

Oct 21 19:58:48 scylla-scorecard-a-n01 scylla[17426]:  [shard 0] rpc - client 52.26.187.180:7001: fail to connect: Connection timed out

n15:

$ cat n15 |grep on_dead
Oct 22 17:13:09 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] storage_service - endpoint=54.198.116.4 on_dead
Oct 22 19:10:23 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - stream_manager: Close all stream_session with peer = 3.82.217.32 in on_dead
Oct 22 19:18:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] storage_service - endpoint=34.211.128.153 on_dead
Oct 22 19:23:05 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - stream_manager: Close all stream_session with peer = 54.157.198.151 in on_dead
```

On n15 (scylla-scorecard-a-n15.use1.log), I saw a lot of schema_registry messages.
```
   5752 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5753 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
   5754 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked af38d702-4d3d-3620-91e4-016f279d0664 as synced
   5755 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
   5756 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5757 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
   5758 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked af38d702-4d3d-3620-91e4-016f279d0664 as synced
   5759 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
   5760 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
   5761 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
   5762 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5763 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 1] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
   5764 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5765 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5766 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
   5767 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5768 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
   5769 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
```
@tgrabiec why did we see such schema_registry marked as synced again and again?


And I saw

```
   3920 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{0007636c69636b6564},0}
   3921 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3922 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3923 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3924 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3925 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3926 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3927 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3928 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3929 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{00066f70656e6564},0}
   3930 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
```

and

````
3958 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3959 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3960 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3961 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3962 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3963 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3964 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3965 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3966 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3967 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{0007636c69636b6564},0}
   3968 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{00066f70656e6564},0}
   3969 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3970 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
   3971 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] clustering row:{position: clustered,ckp{000964656c697665726564},0}
   3972 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition end:{position: partition_end,null,0}
   3973 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 6] flat_mutation_reader - [validator 0x6060140ffef0] partition start:{position: partition_start,null,0}
```

@denesb what does the validator complain?

**Comment by asias** [comment]
My theory is bootstrapping node (n15) is overwhelmed (Semaphore timed out: smp_service_group), and existing node that sends data to n15, e.g., n1 marked n15 as down, as a result the streaming session for bootstrap failed. There were 24 nodes in the cluster. We allow stream from 16 nodes in parallel at most. 

```
183     // Limit the number of nodes to stream in parallel to reduce memory pressure with large cluster.
184     seastar::semaphore _limiter{16};
```

n01:
```
 462 Oct 22 15:36:06 scylla-scorecard-a-n01 scylla[17426]:  [shard 0] gossip - InetAddress 52.55.159.2 is now DOWN, status = BOOT
 463 Oct 22 15:36:09 scylla-scorecard-a-n01 scylla[17426]:  [shard 0] gossip - Fail to send EchoMessage to 52.55.159.2: seastar::rpc::timeout_error (rpc call timed out)
 464 Oct 22 15:36:25 scylla-scorecard-a-n01 scylla[17426]:  [shard 0] gossip - Fail to send EchoMessage to 52.55.159.2: seastar::rpc::timeout_error (rpc call timed out)
```



n15
```
 136082 Oct 22 15:37:42 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] storage_proxy - Failed to apply mutation from 54.202.251.66#10: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#        2 0->10 semaphore)
 136083 Oct 22 15:37:42 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] storage_proxy - Failed to apply mutation from 54.202.251.66#10: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#        2 0->10 semaphore)
 136084 Oct 22 15:37:42 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] storage_proxy - Failed to apply mutation from 54.209.43.160#10: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#        2 0->10 semaphore)
 136085 Oct 22 15:37:42 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] storage_proxy - Failed to apply mutation from 52.39.167.226#10: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#        2 0->10 semaphore)
 136086 Oct 22 15:37:43 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] gossip - InetAddress 34.211.128.153 is now DOWN, status = NORMAL
 136087 Oct 22 15:37:43 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] rpc - client 52.26.187.180:7001: fail to connect: Connection timed out 
 136088 Oct 22 15:37:43 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #8b9764b0-13c2-11eb-8ff4-000000000003] Received failed complete message, peer=54.157.198.151
 136089 Oct 22 15:37:46 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #8b9764b0-13c2-11eb-8ff4-000000000003] Streaming plan for Bootstrap-scorecard-index-9 failed, peers={54.157.198.151        }, tx=0 KiB, 0.00 KiB/s, rx=58278493 KiB, 729.59 KiB/s
 136090 Oct 22 15:37:52 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap with 54.157.198.151 for keyspace=scorecard failed, took 95106.4 seconds: streaming::stream_exception (Stream fail        ed) 
 136091 Oct 22 15:37:56 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] gossip - Fail to send EchoMessage to 34.211.128.153: seastar::rpc::timeout_error (rpc call timed out)
 136092 Oct 22 15:38:07 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] gossip - Fail to send EchoMessage to 34.211.128.153: seastar::rpc::timeout_error (rpc call timed out)
```

**Comment by asias** [comment]
@powellchristoph I would recommend to first fix the node with node failure with the replace operation (--replace_address). Once it is fixed, then we can add the new node.

Let's try adding `--enable_repair_based_node_ops true` for the replacing node and bootstrap node to use the new repair based node operations.

Let me know if this helps. If not, I will figure something else to help you replace and bootstrap.

**Comment by asias** [comment]
@powellchristoph I worked on a patch on to of 4.1 branch for you to make the bootstrap/replace less aggressive.

[scylla-package.scylla-4.1-branch.issue.7476.8690b5bf.tar.gz](https://drive.google.com/file/d/1CDNT2Sm8oURYP14LBuVzds0f_y1XBvBg/view?usp=sharing)

So, try in the following order. 

1) stock 4.1.8  + --enable_repair_based_node_ops true

2) my binary + --enable_repair_based_node_ops false

3 )my binary + --enable_repair_based_node_ops true

To install :
$ tar xvf scylla-package.scylla-4.1-branch.issue.7476.8690b5bf.tar.gz
$ sudo ./install.sh

**Comment by tgrabiec** [comment]
On Wed, Oct 28, 2020 at 2:54 AM Asias He <notifications@github.com> wrote:

> @powellchristoph <https://github.com/powellchristoph> Thanks.
>
> The joining node does not stuck in streaming. The streaming for bootstrap
> failed and it was retried.
>
> 1033377 Oct 22 19:14:18 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839aa000-149a-11eb-8ff4-000000000003] Received failed complete message, peer=3.82.217.32
> 1033378 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839aa000-149a-11eb-8ff4-000000000003] Streaming plan for Bootstrap-scorecard-index-0 failed, peers={3.82.217.32},         tx=0 KiB, 0.00 KiB/s, rx=131015 KiB, 1131.17 KiB/s
> 1033379 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] rpc - client 10.2.96.53:7001: client connection dropped: read: Connection reset by peer
> 1033380 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839bd880-149a-11eb-8ff4-000000000003] Received failed complete message, peer=54.157.198.151
> 1033381 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap with 3.82.217.32 for keyspace=scorecard failed, took 116.496 seconds: streaming::stream_exception (Stream failed)
> 1033382 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - [Stream #839bd880-149a-11eb-8ff4-000000000003] Streaming plan for Bootstrap-scorecard-index-0 failed, peers={54.157.198.151        }, tx=0 KiB, 0.00 KiB/s, rx=136129 KiB, 1170.94 KiB/s
> 1033383 Oct 22 19:14:19 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap with 54.157.198.151 for keyspace=scorecard failed, took 116.957 seconds: streaming::stream_exception (Stream fail        ed)
> 1033384 Oct 22 19:14:20 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap failed, took 134 seconds, nr_ranges_remaining=1377
> 1033385 Oct 22 19:14:20 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] range_streamer - Bootstrap failed to stream. Will retry in 90 seconds ...
>
>
> So n15(52.55.159.2) is the bootstrapping node. On which node did you run
> replace operation? You mentioned you did replace operation. Who were node
> 54.157.198.151 / 52.26.187.180 and 3.82.217.32? What happened to them?
>
>
> n1:
>
> Oct 21 19:58:48 scylla-scorecard-a-n01 scylla[17426]:  [shard 0] rpc - client 52.26.187.180:7001: fail to connect: Connection timed out
>
> n15:
>
> $ cat n15 |grep on_dead
> Oct 22 17:13:09 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] storage_service - endpoint=54.198.116.4 on_dead
> Oct 22 19:10:23 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - stream_manager: Close all stream_session with peer = 3.82.217.32 in on_dead
> Oct 22 19:18:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] storage_service - endpoint=34.211.128.153 on_dead
> Oct 22 19:23:05 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] stream_session - stream_manager: Close all stream_session with peer = 54.157.198.151 in on_dead
>
> On n15 (scylla-scorecard-a-n15.use1.log), I saw a lot of schema_registry
> messages.
>
>    5752 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5753 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
>    5754 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked af38d702-4d3d-3620-91e4-016f279d0664 as synced
>    5755 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 10] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
>    5756 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5757 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
>    5758 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked af38d702-4d3d-3620-91e4-016f279d0664 as synced
>    5759 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
>    5760 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
>    5761 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
>    5762 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5763 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 1] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
>    5764 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5765 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5766 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 7b02d1ea-57c0-37a8-a56b-2b7f124a4a39 as synced
>    5767 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 11] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5768 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 0] schema_registry - Marked 2c8c47c0-1830-3527-a372-33e0b1d787d8 as synced
>    5769 Oct 22 15:17:39 scylla-scorecard-a-n15 scylla[1740]:  [shard 5] schema_registry - Marked 96f7d9d6-3d5e-312e-bf3e-3e6f7d9a7d68 as synced
>
> @tgrabiec <https://github.com/tgrabiec> why did we see such
> schema_registry marked as synced again and again?
>

If this happens during schema pull, it's expected. Does it happen all the
time?

>

**Comment by powellchristoph** [comment]
@asias I will follow your suggestions today.

> You mentioned you did replace operation. 

Yes, it died. I attempted to replace it the same way and it failed in the same manner. It has been offline for a few days now. I will decommission the failed node in the other DC first following your steps.

**Comment by asias** [comment]
> 
> 
> @asias I will follow your suggestions today.
> 
> > You mentioned you did replace operation.
> 
> Yes, it died. I attempted to replace it the same way and it failed in the same manner. It has been offline for a few days now. I will decommission the failed node in the other DC first following your steps.



I recommend not to remove node (either nodetool decommission or nodetool removenode) at the moment. It is best to use replace procedure first to recover the dead node. 

Use:

my binary + --enable_repair_based_node_ops true + --replace_address

to perform the replace operation which is much safer than  --enable_repair_based_node_ops false (by default)

**Comment by powellchristoph** [comment]
@asias I attempted to replace the failed node according to your instructions and it won't start. It runs for a few minutes and does the whole joining bit, and then just shuts down without anything logged.

```
Oct 28 21:31:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] storage_service - JOINING: schema complete, ready to bootstrap
Oct 28 21:31:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] storage_service - JOINING: waiting for pending range calculation
Oct 28 21:31:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] storage_service - JOINING: calculation complete, ready to bootstrap
Oct 28 21:31:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] storage_service - Checking bootstrapping/leaving nodes: ok
Oct 28 21:31:46 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - Fail to send EchoMessage to 52.26.187.180: seastar::rpc::timeout_error (rpc call timed out)
Oct 28 21:32:13 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - FatClient 52.26.187.180 has been silent for 30000ms, removing from gossip
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] init - Shutting down gossiping
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - My status = UNKNOWN
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - No local state or state is in silent shutdown, not announcing shutdown
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - Disable and wait for gossip loop started
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - Disable and wait for gossip loop finished
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] gossip - Gossip is now stopped
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] init - Shutting down gossiping was successful
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] init - Shutting down storage service notifications
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] init - Shutting down storage service notifications was successful
Oct 28 21:32:43 scylla-scorecard-a-n04 scylla[96262]:  [shard 0] init - Shutting down repair message handlers
```

I have run the `node_health_check` on the failing node and another running node in the same region. There is no core dump to provide.

https://drive.google.com/file/d/1itbFmZcoBVNGUjyg8_kWsfpuG5QcZhoa/view?usp=sharing, https://drive.google.com/file/d/1opvhWaPXRHE32WYz5r9QO1DX3-fTHgW2/view?usp=sharing

**Comment by asias** [comment]
@powellchristoph Please provide the full log of the new node that is trying to replace 52.26.187.180.
 
Also provide the output of `nodetool status` and `nodetool gossipinfo ; sleep 5; nodetool gossipinfo ; sleep 3;  nodetool gossipinfo ` on all nodes in the cluster. 

Did you shutdown all the nodes in the cluster once after node 52.26.187.180 is down? I think you probably hit a limitation that a node can not be replaced if the whole cluster was shutdown. I had a merged patch in the past to help this situation: "storage_service: Allow replacing node when all nodes were shutdown", but the patch is not merged. 

I will try to figure out a solution for you to replace 52.26.187.180.

**Comment by asias** [comment]
@powellchristoph I had a workaround to allow "shutdown all nodes and replace".  [Here](https://drive.google.com/file/d/1nV41XiAWVuKQexny-e_nSTkl8oW30JPz/view?usp=sharing) it is: 



If you see log like

```
Cannot replace token {} which does not exist
```

on the replacing node. You can use the above binary to work around it. According to you log, it is safe to use the workaround.

**Comment by asias** [comment]
@powellchristoph 

Btw, you could add 

```
--compaction_enforce_min_threshold true
--compaction_static_shares 50
```
to the replacing node or bootstrap node.

This will reduce the comapction work significant during the node operations.

**Comment by powellchristoph** [comment]
@asias There was no point where the entire cluster was shutdown and I do not see a message similar to what you referenced.

I have attempted to replace that node again with the additional settings and it starts then stops itself.

Here is the information you requested for all nodes in the cluster. 
https://drive.google.com/file/d/1-AX0MFOmeP6YHUnDnl1VQpp69Ti2oJm6/view?usp=sharing, https://drive.google.com/file/d/1RQpdPVlE8wdLoJUWGwJMR8xBLj_pgMtO/view?usp=sharing

**Comment by asias** [comment]
> @asias There was no point where the entire cluster was shutdown and I do not see a message similar to what you referenced.
> 
> I have attempted to replace that node again with the additional settings and it starts then stops itself.
> 
> Here is the information you requested for all nodes in the cluster.
> https://drive.google.com/file/d/1-AX0MFOmeP6YHUnDnl1VQpp69Ti2oJm6/view?usp=sharing, https://drive.google.com/file/d/1RQpdPVlE8wdLoJUWGwJMR8xBLj_pgMtO/view?usp=sharing

I looked at the log. The last log is right before the error message I suspected. Perhaps The throw err is not logged properly. Please use the version I posted [here](https://github.com/scylladb/scylla/issues/7476#issuecomment-718307563). It skipped the check and added log to verify.

```
slogger.info("Checking bootstrapping/leaving nodes: ok");
...
throw std::runtime_error(format("Cannot replace token {} which does not exist!", token));
```

**Comment by powellchristoph** [comment]
@asias Its joining now with the version you provided and all suggested settings.

Here is the existing information from it currently as its joining:
https://drive.google.com/file/d/12KLRZ9H9BL6FxViNgt8OzHtzxPqI1hni/view?usp=sharing

**Comment by asias** [comment]
> @asias Its joining now with the version you provided and all suggested settings.
> 
> Here is the existing information from it currently as its joining:
> https://drive.google.com/file/d/12KLRZ9H9BL6FxViNgt8OzHtzxPqI1hni/view?usp=sharing

Cheers. Let me know if it manages to replace.

**Comment by powellchristoph** [comment]
@asias Success!! Should I continue to use this build and/or settings to continue to expand my cluster?

```
[prod-us-west-2]cpowell@scylla-scorecard-a-n04:~$ nodetool netstats
Mode: NORMAL
Not sending any streams.
Read Repair Statistics:
Attempted: 0
Mismatch (Blocking): 0
Mismatch (Background): 0
Pool Name                    Active   Pending      Completed   Dropped
Large messages                  n/a         0      865692683         0
Small messages                  n/a         0              0         0
Gossip messages                 n/a         0              0         0
```

**Comment by asias** [comment]
> @asias Success!! Should I continue to use this build and/or settings to continue to expand my cluster?

Nice!

Yes, please use my binary + the options https://github.com/scylladb/scylla/issues/7476#issuecomment-718454385 + --enable_repair_based_node_ops true

to add the new node. 

Keep me posted.

Btw, with enable_repair_based_node_ops true, scylla does not use the old streaming to sync data. So `nodetool netstats` will not be relevant.

**Comment by powellchristoph** [comment]
Are these settings required permanently or just simply until the node `JOIN`s successfully? I am currently hand building these machines, rather than in my config managment. Are these settings that I should use everywhere?

**Comment by powellchristoph** [comment]
@asias The replacement node appears to be running properly. How do I progress moving forward?

**Comment by asias** [comment]
Once the bootstrap or replace operation is finished on the new node, you can just switch back to the normal binary and drop the extra options. 

I mentioned here: https://github.com/scylladb/scylla/issues/7476#issuecomment-720070053. You can continue to add new new if you want. Or you already have managed to add another new node after the replace operation?

**Comment by powellchristoph** [comment]
@asias I have replaced the failed node and scaled out with an additional node.

**Comment by avikivity** [comment]
@asias @powellchristoph what's the status? is there still an operational problem? do we need to fix something?

**Comment by asias** [comment]
I think @powellchristoph managed to complete the replace and bootstrap operation. 

To improve scylla, I have an idea of introducing a flag --conservative-node-ops which make the node operations less aggressive to help the case where users have problems to add or remove a node, especially add a node.  For example we can  reduce the number of nodes we stream from in parallel (streaming based node ops, hardcoded to 16),  number of vnode ranges we work in parallel (repair based node ops, hardcoded to 3 ranges per 1GB shard memory). It also sets options like: --compaction_enforce_min_threshold true, --compaction_static_shares 50 automatically to reduce compaction work.

This flag is easy for user to use. The downside is that it can not control individual flags independently. Perhaps it is good enough to have: --conservative-node-ops 0.5 which means it is 50% as aggressive as default. If user sets the individual flag, we respect it.

**Comment by avikivity** [comment]
First, those flags are never easy since they require editing configuration files for a single operation. They should be on the nodetool command, not configuration.

I propose this:
 - if there is no seed node configuration, the node boots into a waiting state (maintenance mode)
 - you then issue the command `nodetool boostrap --initial` for the first node or `nodetool bootsrtrap --contact` for the following nodes
 - you can add other parameters to the bootstrap command

But, I don't understand why streaming more slowly will help.

**Comment by slivne** [comment]
@asias / @avikivity - if we are discussing enhancements lets separate this to a new issue.

I am closing this issue since the user has been able to bootstrap - with some additional manual steps
